src/sync.h

```diff
@@ -93,73 +95,72 @@ void static inline DeleteLock(void* cs) {}
 #endif
 #define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
 
-/**
- * Wrapped mutex: supports recursive locking, but no waiting
- * TODO: We should move away from using the recursive lock by default.
- */
-class CCriticalSection : public AnnotatedMixin<std::recursive_mutex>
+template <typename Mutex>
+class LockOrderMixin : public Mutex
 {
 public:
-    ~CCriticalSection() {
+    ~LockOrderMixin() {
```

Comment:
Maybe just move this into AnnotatedMixin instead of making everything two classes on top of each other?